pageserver/src/tenant/remote_timeline_client/download.rs

```diff
@@ -73,33 +74,129 @@ pub async fn download_layer_file<'a>(
+                                format!("create a destination file for layer '{temp_file_path}'")
+                            })
+                            .map_err(DownloadError::Other)?;
+
+                        let mut download = storage.download(&remote_path, cancel).await?;
+
+                        // TODO: use vectored write (writev) once supported by tokio-epoll-uring.
+                        // There's chunks_vectored() on the stream.
+                        let (bytes_amount, destination_file) = async {
+                            let size_tracking = size_tracking_writer::Writer::new(destination_file);
+                            let mut buffered = owned_buffers_io::write::BufferedWriter::<
+                                { super::BUFFER_SIZE },
+                                _,
+                            >::new(size_tracking);
+                            while let Some(res) =
+                                futures::StreamExt::next(&mut download.download_stream).await
+                            {
+                                let chunk = match res {
+                                    Ok(chunk) => chunk,
+                                    Err(e) => return Err(e),
+                                };
+                                buffered
+                                    .write_buffered(tokio_epoll_uring::BoundedBuf::slice_full(
+                                        chunk,
+                                    ))
+                                    .await?;
+                            }
+                            let size_tracking = buffered.flush_and_into_inner().await?;
+                            Ok(size_tracking.into_inner())
+                        }
+                        .await?;
+
+                        // not using sync_data because it can lose file size update
+                        destination_file
+                            .sync_all()
+                            .await
+                            .with_context(|| {
+                                format!("failed to fsync source file at {temp_file_path}")
+                            })
+                            .map_err(DownloadError::Other)?;
+
+                        Ok(bytes_amount)
+                    }
+                    .await
                 }
+            };
+
+            // in case the download failed, clean up
+            match res {
+                Ok(bytes_amount) => Ok(bytes_amount),
                 Err(e) => {
                     if let Err(e) = tokio::fs::remove_file(&temp_file_path).await {
-                        on_fatal_io_error(&e, &format!("Removing temporary file {temp_file_path}"));
+                        if e.kind() != std::io::ErrorKind::NotFound {
+                            on_fatal_io_error(
+                                &e,
+                                &format!("Removing temporary file {temp_file_path}"),
+                            );
+                        }
```

Comment:
Just highlighting this drive-by-fix, see PR description

> - if we failed in retry_download to create a file, we'd try to remove it, encounter `NotFound`, and `abort()` the process using `on_fatal_io_error`. This PR adds treats `NotFound` as a success.

Reply:
Thanks for highlighting.